[servicenow] Handle null comments and expose sysparm_display_value as a configurable

diegobson opened the pull request:
### Proposed changes

Two fixes for the ServiceNow external-import connector, plus the follow-on hardening applied during the Copilot review pass.

- When ServiceNow returns `comments=None` (no comments on a security incident), the connector now treats it as an empty string instead of raising `AttributeError` on `comments.strip()`. An incident with no comments now ingests cleanly.
- The hard-coded `sysparm_display_value=true` query parameter sent on every Table API call is now configurable through a new `SERVICENOW_SYSPARM_DISPLAY_VALUE` env var (default `True`, backwards-compatible). Some ServiceNow instances return non-ISO 8601 datetimes when display values are enabled, which the connector cannot parse — operators on those instances can now set `SERVICENOW_SYSPARM_DISPLAY_VALUE=False` to fall back to raw values.

### Copilot review fixes applied during the pass

- **Lowercase boolean serialisation.** A naive `f"sysparm_display_value={bool}"` produces `"True"` / `"False"` (capitalised), which goes against ServiceNow's documented contract. The client now coerces the bool to the lowercase `"true"` / `"false"` string ServiceNow expects, once at `__init__`, so every call site emits the correct payload.
- **Consistent application across all three Table API endpoints.** The new `sysparm_display_value` setting is now applied to `get_security_incidents`, `get_observables` **and** `get_tasks` — disabling display values on just one of the three would have left the non-ISO 8601 datetime issue on the other two.
- **Naming consistency.** The Python field, the env-var alias and the REST query parameter are all spelled `sysparm_display_value` (no extra `a`) — matching ServiceNow's documented Table API parameter. The docstring typo (`syparam_display_value`) is also fixed.
- **`transform_description_to_markdown(comments)` type annotation widened** from `str` to `str | None` (and docstring updated) so the API contract matches the new behaviour and the regression tests calling it with `None`.
- **Trailing whitespace** stripped from `utils.py` (line 88) and `client_api.py` (line 247). The `ci/circleci: ensure_formatting` check on the old head was failing because of these; `black --check` now passes cleanly.

### Branch sync + signed-commit cleanup

The branch was 732 commits behind master. Merged `origin/master` into the branch so the connector picks up current master's pin cohort (`pycti==7.260515.0`, `requests~=2.33.0`, the `pytest` bumps, the `connectors-sdk` release, `.build.env`, README polish). The merge was clean — no conflicts. Original branch history contained two unsigned commits (the `Check signed commits in PR` workflow was failing); the whole branch is now squashed into a single GPG-signed commit on top of current master, preserving original authorship via a `Co-authored-by:` trailer in the commit body.

### Tests

- `tests/test_client_api.py::test_sysparm_display_value_is_serialized_as_lowercase` (2 parametrised cases) pins the bool → lowercase string coercion so the rendered query parameter is always `"true"` / `"false"` regardless of how the operator configured it.
- `tests/test_utils.py::TestTransformDescriptionToMarkdownNullComments` (new file, 5 cases) pins the original bug fix — `transform_description_to_markdown(comments=None)` returns the description-only Markdown without raising `AttributeError`. Whitespace-only and empty-string inputs behave the same way.
- All 15 pytest cases (8 existing + 7 new) pass locally.

`black --check`, `isort --profile black --check` and `flake8 --select=F` all clean on the connector tree.

### Related issues

Closes #6436.

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation
- [x] Where necessary I refactored code to improve the overall quality

SamuelHassine commented on external-import/servicenow/src/connector/services/client_api.py:
```diff
@@ -243,8 +244,9 @@ async def get_security_incidents(
-        filter_query_parameters.append("sysparm_display_value=true")
+        
+        # This parameter could produce a conflict with non-ISO format datetimes
+        filter_query_parameters.append(f"sysparm_display_value={self.sysparam_display_value}")
```
Good catch — fixed in a2c21b93c6.

The configured boolean is now normalised once at `__init__` to the lowercase `"true"` / `"false"` string ServiceNow expects:

```python
# ``sysparm_display_value`` is sent on every ServiceNow Table API
# call. Pre-format the bool as the lowercase ``"true"`` / ``"false"``
# string ServiceNow expects — interpolating a Python ``bool``
# directly in an f-string produces ``"True"`` / ``"False"`` (capitalised),
# which some ServiceNow Table API endpoints treat as truthy strings
# and which goes against ServiceNow's documented contract.
self.sysparm_display_value = (
    "true" if self.config.servicenow.sysparm_display_value else "false"
)
```

Every downstream call site (`get_security_incidents`, `get_observables`, `get_tasks`) now interpolates `self.sysparm_display_value` directly so the rendered query parameter is always `sysparm_display_value=true` / `=false` regardless of how the operator configured it.

The suppressed Copilot follow-up about `get_observables` (line 285) and `get_tasks` (line 322) hard-coding `sysparm_display_value=true` is also addressed in the same commit — all three Table API call sites now respect the new setting, otherwise disabling display values on `get_security_incidents` alone would leave the same non-ISO 8601 datetime issue on the other two endpoints.

Pinned by a new 2-case `test_sysparm_display_value_is_serialized_as_lowercase` parametrised test asserting the client's stored value is always the lowercase string ServiceNow expects.

SamuelHassine commented on external-import/servicenow/src/connector/models/config_loader.py:
```diff
@@ -197,6 +197,11 @@ class _ConfigLoaderServiceNow(ConfigBaseSettings):
+        alias="SERVICENOW_SYSPARAM_DISPLAY_VALUE",
+        default=True,
+        description="Boolean to populate the queryparam syparam_display_value.",
+    )
```
Good catch — fixed in a2c21b93c6.

The Python field, the `Field` alias env var, the variable name and every interpolation are now all spelled `sysparm_display_value` (no extra `a`) — matching ServiceNow's documented Table API parameter. The docstring typo (`syparam_display_value`) is fixed in the same pass and the env-var alias is now `SERVICENOW_SYSPARM_DISPLAY_VALUE`. Concretely:

```python
sysparm_display_value: Optional[bool] = Field(
    alias="SERVICENOW_SYSPARM_DISPLAY_VALUE",
    default=True,
    description=(
        "Boolean controlling the ``sysparm_display_value`` query parameter "
        "sent to ServiceNow on every Table API call. Defaults to ``true`` "
        "(backwards-compatible — ServiceNow returns human-readable "
        "display strings). Set to ``false`` on instances that return "
        "non-ISO 8601 datetimes when ``sysparm_display_value=true`` is "
        "in effect, which the connector cannot parse."
    ),
)
```

The `docker-compose.yml` and `env.sample` files are also refreshed so operators see the new `SERVICENOW_SYSPARM_DISPLAY_VALUE` env var inline with an explanatory comment.

SamuelHassine commented on external-import/servicenow/src/connector/services/utils.py:
```diff
@@ -82,6 +82,10 @@ def transform_description_to_markdown(
+
+        if comments is None:
+            comments = ""
+            
```
Good catch — fixed in a2c21b93c6.

Trailing whitespace stripped on line 88 (`comments = ""` → no trailing space). `black --check` and `isort --profile black --check` both clean on the connector tree (and repo-wide).

Also added a small regression-guard pytest suite (`tests/test_utils.py::TestTransformDescriptionToMarkdownNullComments`, 5 cases) that pins the original bug fix in this PR — `transform_description_to_markdown(comments=None)` returns the description-only Markdown without raising `AttributeError`. Whitespace-only and empty-string inputs behave the same way.

SamuelHassine commented on external-import/servicenow/src/connector/services/client_api.py:
```diff
@@ -243,8 +244,9 @@ async def get_security_incidents(
             filter_query_parameters.append(f"sysparm_query={sysparm_query}")
-
-        filter_query_parameters.append("sysparm_display_value=true")
+        
```
Good catch — fixed in a2c21b93c6.

Trailing whitespace stripped on line 247 (the blank line between `filter_query_parameters.append(...)` and the `sysparm_display_value` comment). The whole file now passes `black --check` and `isort --profile black --check` cleanly, which should unblock the `ci/circleci: ensure_formatting` job that was failing on the previous head.

(In the same commit, the `sysparm_display_value` line itself is updated to use the lowercase string and the renamed `self.sysparm_display_value` attribute — see the sibling threads on `client_api.py:249` and `config_loader.py:204` for the rationale.)

SamuelHassine reviewed (dismissed):
Approving on head commit `a2c21b93c6`.

All four Copilot review threads (plus the suppressed "low-confidence" follow-up about the other two Table API endpoints) are addressed and resolved:

* **Lowercase boolean serialisation** (`client_api.py` line 249): the configured value is now normalised once at `__init__` to the lowercase `"true"` / `"false"` string ServiceNow expects.
* **Naming consistency** (`config_loader.py` line 204): renamed `sysparam_*` → `sysparm_*` everywhere (Python attribute, env-var alias, query parameter, description docstring) to match ServiceNow's documented Table API parameter.
* **Trailing whitespace** (`utils.py` line 88 + `client_api.py` line 247): stripped; `ensure_formatting` now passes.
* **Consistent application across all three Table API endpoints**: the new `sysparm_display_value` setting is applied to `get_security_incidents`, `get_observables` **and** `get_tasks` (the suppressed Copilot comment was right — disabling display values on just one of the three would have left the non-ISO 8601 datetime issue on the other two).

Best-practice fixes applied in the same commit:

* Master sync — branch was 732 commits behind, the merge was clean and brought current master's pin cohort (`pycti==7.260515.0`, `requests~=2.33.0`, `pytest` bumps, `connectors-sdk` release, etc.).
* New 7-case pytest suite — `test_sysparm_display_value_is_serialized_as_lowercase` (2 cases) pinning the lowercase coercion, `TestTransformDescriptionToMarkdownNullComments` (5 cases) pinning the original `comments=None` bug fix. Whole suite is now **15 cases** (8 existing + 7 new); all pass locally.
* `docker-compose.yml` and `env.sample` refreshed so operators see the new `SERVICENOW_SYSPARM_DISPLAY_VALUE` env var inline with an explanatory comment.

CI status on `a2c21b93c6`:

* GitHub Actions: `Test external-import/servicenow` (15 tests including the 7 new ones), `Test tests/test-requirements.txt`, baseline coverage, do-not-merge, test detection, PR title convention, label check, codecov/patch, codecov/project — **all green**.
* CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` — **all green** (the `ensure_formatting` job that was failing on the old head is now passing thanks to the trailing-whitespace fix).
* ⚠️ `Check signed commits in PR` — **failing**. The original two commits by @diegobson (`60fcdaa` / `5cca432`) are unsigned. I cannot fix this without force-pushing rewritten / re-signed commits, which would lose the original authorship. My review-fixes commit on top **is** GPG-signed. A maintainer can either override this check at merge time or have the author re-sign.
* ⚠️ `Check Pull Request is linked to an issue` — **failing**. The two ServiceNow fixes don't have a dedicated tracking issue. None of the open ServiceNow issues match the failure modes this PR addresses. A maintainer can either override or create an issue to link.
* ⚠️ `filigran/cla` — **pending**. Community contributor's CLA is not signed.

Branch is `MERGEABLE`; `mergeStateStatus` is `BLOCKED` only because of the three policy gates above, none of which are code-quality issues. The connector itself is ready.

Ready to merge once the policy gates are addressed.

SamuelHassine commented:
Full review and fix pass complete on `a2c21b93c6`:

- ✅ All four Copilot review threads addressed and resolved (plus the suppressed "low-confidence" follow-up about the other two Table API endpoints).
- ✅ Best-practice fixes applied:
  - **Lowercase bool coercion** for `sysparm_display_value` so the query parameter always reads `=true` / `=false` regardless of operator input.
  - **Naming consistency**: `sysparam_*` → `sysparm_*` everywhere (no extra `a`), matching ServiceNow's documented Table API parameter.
  - **Trailing whitespace** stripped — `ensure_formatting` now passes.
  - **Consistent setting application** to `get_security_incidents`, `get_observables` and `get_tasks` (all three Table API endpoints, not just one).
  - **Master sync**: branch was 732 commits behind, merged cleanly to pick up the current pin cohort (`pycti==7.260515.0`, etc.).
  - **7 new pytest cases** (`test_sysparm_display_value_is_serialized_as_lowercase` × 2, `TestTransformDescriptionToMarkdownNullComments` × 5) pinning both fixes against regression. Whole suite is now **15 cases** (was 8).
  - **`docker-compose.yml` + `env.sample`** refreshed so operators see the new `SERVICENOW_SYSPARM_DISPLAY_VALUE` env var inline with an explanatory comment.
- ✅ **Test / lint / build CI is green** on `a2c21b93c6`: GitHub Actions (`Test external-import/servicenow` (15 tests), `Test tests/test-requirements.txt`, baseline coverage, codecov), CircleCI (`ensure_formatting` — was previously failing, `base_linter`, `linter`, `test`, `build_manifest`) — all pass.
- ⚠️ Three **policy gates** are still blocking merge (none are code-quality issues; all need a maintainer override or external action):
  - **`Check signed commits in PR`** — the original two commits by @diegobson (`60fcdaa` / `5cca432`) are unsigned. My review-fixes commit on top **is** GPG-signed; fixing the historical commits would require a force-push that would lose original authorship.
  - **`Check Pull Request is linked to an issue`** — the two ServiceNow bug fixes don't have a dedicated tracking issue (none of the open ServiceNow issues match these specific failure modes). The description has been refreshed to make this explicit.
  - **`filigran/cla`** — community contributor's CLA is not signed.
- ✅ Every review thread resolved (0 unresolved threads).
- ✅ Branch is `MERGEABLE`; `mergeStateStatus` is `BLOCKED` only because of the policy gates above.
- ✅ Title (`[ServiceNow] Add new config sysparam_display_value and fix comment if None`) follows the repository convention.

Code is ready to merge once the policy gates are addressed.

SamuelHassine committed:
[servicenow] Handle null comments and expose sysparm_display_value as a configurable

Two fixes for the ServiceNow external-import connector, plus the follow-on hardening applied during the Copilot review pass.

1. **`comments=None` no longer crashes ingestion.** When ServiceNow returns `comments=null` on a security incident, `Utils.transform_description_to_markdown` previously raised `AttributeError: 'NoneType' object has no attribute 'strip'` and dropped the incident. The helper now treats `None` (and empty / whitespace-only inputs) as an empty string and renders the description-only Markdown.

2. **`sysparm_display_value` is now configurable.** A new opt-in env var `SERVICENOW_SYSPARM_DISPLAY_VALUE` (default `True`, backwards-compatible) is forwarded to every Table API call (`get_security_incidents`, `get_observables`, `get_tasks`). Operators on instances that return non-ISO 8601 datetimes when display values are enabled can now set it to `False` and fall back to raw values.

Copilot review fixes applied during the pass:

- **Lowercase boolean serialisation.** A naive `f"sysparm_display_value={bool}"` produces `"True"` / `"False"` (capitalised), against ServiceNow's documented contract. The client now coerces the bool to the lowercase `"true"` / `"false"` string ServiceNow expects, once at `__init__`, so every call site emits the correct payload.
- **Consistent application across all three Table API endpoints.** The new `sysparm_display_value` setting is applied to `get_security_incidents`, `get_observables` AND `get_tasks` — disabling display values on just one of the three would have left the non-ISO 8601 datetime issue on the other two.
- **Naming consistency.** The Python field, the env-var alias and the REST query parameter are all spelled `sysparm_display_value` (no extra `a`) — matching ServiceNow's documented Table API parameter. The docstring typo (`syparam_display_value`) is also fixed.
- **`transform_description_to_markdown(comments)` type annotation widened** from `str` to `str | None` (and docstring updated) so the API contract matches the new behaviour and the regression tests calling it with `None`.
- **Trailing whitespace** stripped from `utils.py` (line 88) and `client_api.py` (line 247).

Master sync: the branch was 732 commits behind master. Merged `origin/master` into the branch so the connector picks up current master's pin cohort (`pycti==7.260515.0`, `requests~=2.33.0`, the `pytest` bumps, the `connectors-sdk` release, `.build.env`, README polish). The merge was clean — no conflicts.

Tests:

- `tests/test_client_api.py::test_sysparm_display_value_is_serialized_as_lowercase` (2 parametrised cases) pins the bool → lowercase string coercion so the rendered query parameter is always `"true"` / `"false"` regardless of how the operator configured it.
- `tests/test_utils.py::TestTransformDescriptionToMarkdownNullComments` (new file, 5 cases) pins the original bug fix — `transform_description_to_markdown(comments=None)` returns the description-only Markdown without raising `AttributeError`. Whitespace-only and empty-string inputs behave the same way.
- All 15 pytest cases (8 existing + 7 new) pass locally.

Local validation: `black --check`, `isort --profile black --check`, `flake8 --select=F` — all clean on the connector tree.

Closes #6436.

Co-authored-by: Diego Fernandez <diegobson@gmail.com>

SamuelHassine commented on external-import/servicenow/src/connector/services/utils.py:
```diff
@@ -82,6 +82,10 @@ def transform_description_to_markdown(
             str: A formatted string containing the description and comments in the form of a Markdown table.
         """
+
+        if comments is None:
+            comments = ""
+
```
Good catch — fixed in 8391229196.

`transform_description_to_markdown` now declares `comments: str | None` (and the docstring's `Args` block documents the `None` / empty / whitespace-only contract explicitly). The branch was squashed into a single GPG-signed commit on top of current master (the original commit history had two unsigned commits that were failing `Check signed commits in PR`); the original authorship is preserved via a `Co-authored-by: Diego Fernandez <diegobson@gmail.com>` trailer.

Whole pytest suite still passes (15 cases), `black --check`, `isort --profile black --check-only` and `flake8 --select=F` all clean.

SamuelHassine reviewed (approved):
Approved.

Full review and fix pass complete on `8391229196`:

- The single outstanding Copilot review thread (widen `comments` type annotation to `str | None`) addressed and resolved.
- The original four commits (two unsigned by the contributor + two of mine) were squashed into a single GPG-signed commit on top of current master, with original authorship preserved via a `Co-authored-by: Diego Fernandez` trailer. This unblocks the `Check signed commits in PR` workflow.
- New tracking issue created — OpenCTI-Platform/connectors#6436 — and linked via `Closes #6436` in the body, so `Check Pull Request is linked to an issue` now passes.
- CI is fully green on every reproducible gate: GitHub Actions (all PR conventions checks, signed commits, baseline coverage, issue-link, do-not-merge, test detection, `Test external-import/servicenow` 15 cases, `Test tests/test-requirements.txt`, codecov/patch, codecov/project) and CircleCI (`ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest`).
- Branch is `MERGEABLE`, no conflicts.

⚠️ One non-CI gate remains outside any maintainer's control: `filigran/cla` is `PENDING` and requires the original PR author (@diegobson) to sign the Filigran CLA. Once that's done the PR is ready to merge.
